README.md
# TimeSinceTextView

This is a subclass of android.widget.TextView that exposes a method `setDate()` which accepts a `long` Unix timestamp or `java.util.Date`. The view converts the date into a String which describes the date in terms of time since that timestamp. For example, if the current timestamp is Unix 1453503166 and we call `timeSinceTextView.setDate(1453503116)`, "50 seconds ago" is displayed.

[Changelog](CHANGELOG.md)

[Javadoc](https://jitpack.io/com/github/damien5314/TimeSinceTextView/1.2.3/javadoc/)

## Comparison to DateUtils.getRelativeTimeSpanString

I actually wrote this library before I knew about [DateUtils.getRelativeTimeSpanString](http://developer.android.com/reference/android/text/format/DateUtils.html#getRelativeDateTimeString), but the output is actually quite a bit different. The DateUtils implementation should return localized text and allows for customizable flags. See [here](Comparison.md) for a comparison of the output of different time stamps.

## Usage

Simply declare a `TimeSinceTextView` in XML or create one in code.

```xml
<com.ddiehl.timesincetextview.TimeSinceTextView
  android:id="@+id/timestamp"
  android:layout_width="wrap_content"
  android:layout_height="wrap_content" />
```

Then call `setDate(Date)` or `setDate(long)` with a Unix timestamp, and the text will be automatically generated and set to the view.

```java
((TimeSinceTextView) findViewById(R.id.timestamp)).setDate(1452827942);
```

To get an abbreviated form of the converted text, add `app:abbreviated="true"` to your XML layout.

```xml
<com.ddiehl.timesincetextview.TimeSinceTextView
  xmlns:app="http://schemas.android.com/apk/res-auto"
  android:layout_width="wrap_content"
  android:layout_height="wrap_content"
  app:tstv_abbreviated="true" />
```

The class `TimeSince` also contains static methods which can be used to retrieve a relative timestamp string without an instance of `TimeSinceTextView`.

## Add to your project

[![Release](https://jitpack.io/v/damien5314/TimeSinceTextView.svg)](https://jitpack.io/#damien5314/TimeSinceTextView)

```gradle
repositories {
    maven { url "https://jitpack.io" }
}
dependencies {
  compile 'com.github.damien5314:TimeSinceTextView:1.+'
}
```

![Screenshot](/screenshots/1453502946.png)

## Contributions

Pull requests are welcome, in particular in would be nice to have [strings.xml](timesincetextview/src/main/res/values/strings.xml) translated into as many languages as possible.


CHANGELOG.md
Change Log
==========

Version 1.2.3 *(2016-03-29)*
----------------------------------
* Nothing, Jitpack won't build Version 1.2.2 for some reason

Version 1.2.2 *(2016-03-29)*
----------------------------------
* Removed abbreviations from Japanese translation because it's already short enough

Version 1.2.1 *(2016-03-29)*
----------------------------------
* Renamed string resources to avoid clashing in other libraries

Version 1.2.0 *(2016-03-29)*
----------------------------------
* Added Spanish translation (courtesy of @JorgeDLS)
* Added Japanese translation
* **BREAKING** Renamed `abbreviated` attribute to `tstv_abbreviated` in order to prevent clashing with other libraries. This is not backwards compatible.
* Added methods `setAbbreviated`, `isAbbreviated`, and `getDate`.
* Static utility methods in `TimeSinceTextView` are deprecated in favor those in `TimeSince`.
* Added source code to distribution .aar (how did no one ask for this?)
* Added Javadoc to distribution .aar
* Added Apache 2.0 license, not that I really care what you do with this code.

Version 1.1.1 *(2016-02-07)*
----------------------------------
* Added Russian translation (courtesy of @vkotovv)

Version 1.1.0 *(2016-01-27)*
----------------------------------
* Added Simplified Chinese translation (thanks @jewelzqiu!)
* Switched from simple string resources to plural resources which can be more accurately localized in all languages.
* Switched from using long to int in timespan calculation which should improve memory footprint.

Version 1.0.1 *(2016-01-23)*
----------------------------------
* Exposed static method in TimeSinceTextView class for converting a timestamp to String without using the TextView directly.

Version 1.0.0 *(2016-01-22)*
----------------------------------
* Initial release


timesincetextview/src/test/java/com/ddiehl/timesincetextview/TimeSinceTextViewTest.java
package com.ddiehl.timesincetextview;

import android.content.Context;

import org.hamcrest.Matcher;
import org.hamcrest.core.Is;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.runners.MockitoJUnitRunner;

import java.util.Date;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class TimeSinceTextViewTest {
  private static final String TIMESPAN_NOW = "just now";
  private static final String TIMESPAN_YEARS = "%s years ago";
  private static final String TIMESPAN_MONTHS = "%s months ago";
  private static final String TIMESPAN_WEEKS = "%s weeks ago";
  private static final String TIMESPAN_DAYS = "%s days ago";
  private static final String TIMESPAN_HOURS = "%s hours ago";
  private static final String TIMESPAN_MINUTES = "%s minutes ago";
  private static final String TIMESPAN_SECONDS = "%s seconds ago";
  private static final String TIMESPAN_YEAR = "%s year ago";
  private static final String TIMESPAN_MONTH = "%s month ago";
  private static final String TIMESPAN_WEEK = "%s week ago";
  private static final String TIMESPAN_DAY = "%s day ago";
  private static final String TIMESPAN_HOUR = "%s hour ago";
  private static final String TIMESPAN_MINUTE = "%s minute ago";
  private static final String TIMESPAN_SECOND = "%s second ago";
  private static final String TIMESPAN_YEARS_ABBR = "%syr";
  private static final String TIMESPAN_MONTHS_ABBR = "%smo";
  private static final String TIMESPAN_WEEKS_ABBR = "%swk";
  private static final String TIMESPAN_DAYS_ABBR = "%sd";
  private static final String TIMESPAN_HOURS_ABBR = "%sh";
  private static final String TIMESPAN_MINUTES_ABBR = "%sm";
  private static final String TIMESPAN_SECONDS_ABBR = "%ss";

  @Mock private Context mContext;
  private final long mTime = new Date().getTime() / 1000;

  @Before
  public void setUp() {
    when(mContext.getString(R.string.timespan_now))
        .thenReturn(TIMESPAN_NOW);
    when(mContext.getString(R.string.timespan_years))
        .thenReturn(TIMESPAN_YEAR);
    when(mContext.getString(R.string.timespan_months))
        .thenReturn(TIMESPAN_MONTH);
    when(mContext.getString(R.string.timespan_weeks))
        .thenReturn(TIMESPAN_WEEK);
    when(mContext.getString(R.string.timespan_days))
        .thenReturn(TIMESPAN_DAY);
    when(mContext.getString(R.string.timespan_hours))
        .thenReturn(TIMESPAN_HOUR);
    when(mContext.getString(R.string.timespan_minutes))
        .thenReturn(TIMESPAN_MINUTE);
    when(mContext.getString(R.string.timespan_seconds))
        .thenReturn(TIMESPAN_SECOND);
    when(mContext.getString(R.string.timespan_years_plural))
        .thenReturn(TIMESPAN_YEARS);
    when(mContext.getString(R.string.timespan_months_plural))
        .thenReturn(TIMESPAN_MONTHS);
    when(mContext.getString(R.string.timespan_weeks_plural))
        .thenReturn(TIMESPAN_WEEKS);
    when(mContext.getString(R.string.timespan_days_plural))
        .thenReturn(TIMESPAN_DAYS);
    when(mContext.getString(R.string.timespan_hours_plural))
        .thenReturn(TIMESPAN_HOURS);
    when(mContext.getString(R.string.timespan_minutes_plural))
        .thenReturn(TIMESPAN_MINUTES);
    when(mContext.getString(R.string.timespan_seconds_plural))
        .thenReturn(TIMESPAN_SECONDS);
    when(mContext.getString(R.string.timespan_years_abbr))
        .thenReturn(TIMESPAN_YEARS_ABBR);
    when(mContext.getString(R.string.timespan_months_abbr))
        .thenReturn(TIMESPAN_MONTHS_ABBR);
    when(mContext.getString(R.string.timespan_weeks_abbr))
        .thenReturn(TIMESPAN_WEEKS_ABBR);
    when(mContext.getString(R.string.timespan_days_abbr))
        .thenReturn(TIMESPAN_DAYS_ABBR);
    when(mContext.getString(R.string.timespan_hours_abbr))
        .thenReturn(TIMESPAN_HOURS_ABBR);
    when(mContext.getString(R.string.timespan_minutes_abbr))
        .thenReturn(TIMESPAN_MINUTES_ABBR);
    when(mContext.getString(R.string.timespan_seconds_abbr))
        .thenReturn(TIMESPAN_SECONDS_ABBR);
  }

  @After
  public void tearDown() {

  }

  @Test
  public void getFormattedDateString_now() {
    assertThat(
        TimeSinceTextView.getFormattedDateString(
            mTime - 10, mTime, false, mContext),
        is(TIMESPAN_NOW));
  }

  @Test
  public void getFormattedDateString_10seconds() {
    assertThat(
        TimeSinceTextView.getFormattedDateString(
            mTime - 11, mTime, false, mContext),
        is(TIMESPAN_SECONDS, String.valueOf(11)));
  }

  @Test
  public void getFormattedDateString_59seconds() {
    assertThat(
        TimeSinceTextView.getFormattedDateString(
            mTime - 59, mTime, false, mContext),
        is(TIMESPAN_SECONDS, String.valueOf(59)));
  }

  @Test
  public void getFormattedDateString_1minute() {
    assertThat(
        TimeSinceTextView.getFormattedDateString(
            mTime - 1 * 60, mTime, false, mContext),
        is(TIMESPAN_MINUTE, String.valueOf(1)));
  }

  @Test
  public void getFormattedDateString_59minutes() {
    assertThat(
        TimeSinceTextView.getFormattedDateString(
            mTime - 1 * 60 * 60 + 1, mTime, false, mContext),
        is(TIMESPAN_MINUTES, String.valueOf(59)));
  }

  @Test
  public void getFormattedDateString_1hour() {
    assertThat(
        TimeSinceTextView.getFormattedDateString(
            mTime - 1 * 60 * 60, mTime, false, mContext),
        is(TIMESPAN_HOUR, String.valueOf(1)));
  }

  @Test
  public void getFormattedDateString_23hours() {
    assertThat(
        TimeSinceTextView.getFormattedDateString(
            mTime - 1 * 24 * 60 * 60 + 1, mTime, false, mContext),
        is(TIMESPAN_HOURS, String.valueOf(23)));
  }

  @Test
  public void getFormattedDateString_1day() {
    assertThat(
        TimeSinceTextView.getFormattedDateString(
            mTime - 1 * 24 * 60 * 60, mTime, false, mContext),
        is(TIMESPAN_DAY, String.valueOf(1)));
  }

  @Test
  public void getFormattedDateString_6days() {
    assertThat(
        TimeSinceTextView.getFormattedDateString(
            mTime - 1 * 7 * 24 * 60 * 60 + 1, mTime, false, mContext),
        is(TIMESPAN_DAYS, String.valueOf(6)));
  }

  @Test
  public void getFormattedDateString_1week() {
    assertThat(
        TimeSinceTextView.getFormattedDateString(
            mTime - 1 * 7 * 24 * 60 * 60, mTime, false, mContext),
        is(TIMESPAN_WEEK, String.valueOf(1)));
  }

  @Test
  public void getFormattedDateString_4weeks() {
    assertThat(
        TimeSinceTextView.getFormattedDateString(
            mTime - 1 * 30 * 24 * 60 * 60 + 1, mTime, false, mContext),
        is(TIMESPAN_WEEKS, String.valueOf(4)));
  }

  @Test
  public void getFormattedDateString_1month() {
    assertThat(
        TimeSinceTextView.getFormattedDateString(
            mTime - 1 * 30 * 24 * 60 * 60, mTime, false, mContext),
        is(TIMESPAN_MONTH, String.valueOf(1)));
  }

  @Test
  public void getFormattedDateString_12months() {
    assertThat(
        TimeSinceTextView.getFormattedDateString(
            mTime - 1 * 365 * 24 * 60 * 60 + 1, mTime, false, mContext),
        is(TIMESPAN_MONTHS, String.valueOf(12)));
  }

  @Test
  public void getFormattedDateString_1year() {
    assertThat(
        TimeSinceTextView.getFormattedDateString(
            mTime - 1 * 365 * 24 * 60 * 60, mTime, false, mContext),
        is(TIMESPAN_YEAR, String.valueOf(1)));
  }

  @Test
  public void getFormattedDateString_30years() {
    assertThat(
        TimeSinceTextView.getFormattedDateString(
            mTime - 30 * 365 * 24 * 60 * 60, mTime, false, mContext),
        is(TIMESPAN_YEARS, String.valueOf(30)));
  }

  @Test
  public void getFormattedDateString_abbrev() {
    assertThat(
        TimeSinceTextView.getFormattedDateString(
            mTime - 59, mTime, true, mContext),
        is(TIMESPAN_SECONDS_ABBR, String.valueOf(59)));
    assertThat(
        TimeSinceTextView.getFormattedDateString(
            mTime - 1 * 60, mTime, true, mContext),
        is(TIMESPAN_MINUTES_ABBR, String.valueOf(1)));
    assertThat(
        TimeSinceTextView.getFormattedDateString(
            mTime - 1 * 60 * 60, mTime, true, mContext),
        is(TIMESPAN_HOURS_ABBR, String.valueOf(1)));
    assertThat(
        TimeSinceTextView.getFormattedDateString(
            mTime - 1 * 24 * 60 * 60, mTime, true, mContext),
        is(TIMESPAN_DAYS_ABBR, String.valueOf(1)));
    assertThat(
        TimeSinceTextView.getFormattedDateString(
            mTime - 1 * 7 * 24 * 60 * 60, mTime, true, mContext),
        is(TIMESPAN_WEEKS_ABBR, String.valueOf(1)));
    assertThat(
        TimeSinceTextView.getFormattedDateString(
            mTime - 1 * 30 * 24 * 60 * 60, mTime, true, mContext),
        is(TIMESPAN_MONTHS_ABBR, String.valueOf(1)));
    assertThat(
        TimeSinceTextView.getFormattedDateString(
            mTime - 1 * 365 * 24 * 60 * 60, mTime, true, mContext),
        is(TIMESPAN_YEARS_ABBR, String.valueOf(1)));
  }

  private static Matcher<String> is(String matcher, String... args) {
    return Is.is(
        String.format(matcher, args));
  }
}

library/src/main/java/com/ddiehl/timesincetextview/TimeSince.java
package com.ddiehl.timesincetextview;

import android.content.Context;

public class TimeSince {
  private static final int[] TIMESPAN_IDS = {
      R.plurals.tstv_timespan_years,
      R.plurals.tstv_timespan_months,
      R.plurals.tstv_timespan_weeks,
      R.plurals.tstv_timespan_days,
      R.plurals.tstv_timespan_hours,
      R.plurals.tstv_timespan_minutes,
      R.plurals.tstv_timespan_seconds
  };

  private static final int[] TIMESPAN_IDS_ABBR = {
      R.plurals.tstv_timespan_years_abbr,
      R.plurals.tstv_timespan_months_abbr,
      R.plurals.tstv_timespan_weeks_abbr,
      R.plurals.tstv_timespan_days_abbr,
      R.plurals.tstv_timespan_hours_abbr,
      R.plurals.tstv_timespan_minutes_abbr,
      R.plurals.tstv_timespan_seconds_abbr
  };

  private static int NOW_THRESHOLD_SECONDS = 10;

  /**
   * Returns a formatted date string with respect to the current time, optionally abbreviated.
   * @param utc Unix timestamp of the start date
   * @param abbreviated True if we should display the date string in abbreviated form, false otherwise
   * @param context {@link android.content.Context} with access to application {@link android.content.res.Resources}
   * @return String formatted in the form of `X days ago` or locale equivalent
   */
  public static String getFormattedDateString(long utc, boolean abbreviated, Context context) {
    long currentTime = System.currentTimeMillis() / 1000;
    return getFormattedDateString(utc, currentTime, abbreviated, context);
  }

  /**
   * Returns a formatted date string with respect to the given time, optionally abbreviated.
   * @param start Unix timestamp of the start date
   * @param end Unix timestamp of the end date
   * @param abbreviated True if we should display the date string in abbreviated form, false otherwise
   * @param context {@link android.content.Context} with access to application {@link android.content.res.Resources}
   * @return String formatted in the form of `X days ago` or locale equivalent
   */
  public static String getFormattedDateString(
      long start, long end, boolean abbreviated, Context context) {
    int seconds = (int) (end - start);
    int[] units = new int[] {
        seconds / 31536000, // years
        seconds / 2592000, // months
        seconds / 604800, // weeks
        seconds / 86400, // days
        seconds / 3600, // hours
        seconds / 60, // minutes
        seconds };

    String output = "";
    int unit = 0;
    for (int i = 0; i < units.length; i++) {
      unit = units[i];
      if (unit > 0) {
        output = context.getResources().getQuantityString(
            abbreviated ? TIMESPAN_IDS_ABBR[i] : TIMESPAN_IDS[i],
            unit, unit);
        break;
      }
    }

    if (unit == seconds && seconds <= NOW_THRESHOLD_SECONDS) {
      output = context.getString(R.string.tstv_timespan_now);
    }

    return String.format(output, unit);
  }

  private TimeSince() { }

}


library/src/main/java/com/ddiehl/timesincetextview/TimeSinceTextView.java
package com.ddiehl.timesincetextview;

import android.annotation.TargetApi;
import android.content.Context;
import android.content.res.TypedArray;
import android.os.Build;
import android.util.AttributeSet;
import android.widget.TextView;

import java.util.Date;

public class TimeSinceTextView extends TextView {

  private long mTimestamp = 0;
  private boolean mAbbreviated = false;

  public TimeSinceTextView(Context context) {
    super(context);
    init(context, null, 0, 0);
  }

  public TimeSinceTextView(Context context, AttributeSet attrs) {
    super(context, attrs);
    init(context, attrs, 0, 0);
  }

  public TimeSinceTextView(Context context, AttributeSet attrs, int defStyleAttr) {
    super(context, attrs, defStyleAttr);
    init(context, attrs, defStyleAttr, 0);
  }

  @TargetApi(Build.VERSION_CODES.LOLLIPOP)
  public TimeSinceTextView(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
    super(context, attrs, defStyleAttr, defStyleRes);
    init(context, attrs, defStyleAttr, defStyleRes);
  }

  private void init(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
    if (attrs != null) {
      TypedArray ta = context.getTheme()
          .obtainStyledAttributes(attrs, R.styleable.TimeSinceTextView, 0, 0);
      mAbbreviated = ta.getBoolean(R.styleable.TimeSinceTextView_tstv_abbreviated, false);
    }
  }

  /**
   * @return Unix timestamp associated with this instance, or seconds since January 1, 1970
   */
  public long getDate() {
    return mTimestamp;
  }

  /**
   * Sets the timestamp displayed by this view
   * @param date {@link java.util.Date} from which to calculate time passed
   */
  public void setDate(Date date) {
    setDate(date.getTime() / 1000);
  }

  /**
   * Sets the timestamp displayed by this view
   * @param utc Unix timestamp from which to calculate time passed
   */
  public void setDate(long utc) {
    mTimestamp = utc;
    setText(
        TimeSince.getFormattedDateString(
            utc, mAbbreviated, getContext()));
  }

  /**
   * @return True if view is displaying abbreviated timestamp strings, otherwise false
   */
  public boolean isAbbreviated() {
    return mAbbreviated;
  }

  /**
   * Set to true if the view should display abbreviated timestamp strings
   * @param b True if the view should display abbreviated timestamp strings
   */
  public void isAbbreviated(boolean b) {
    mAbbreviated = b;
    invalidate();
    requestLayout();
  }

  /**
   * @deprecated
   * Use {@link TimeSince#getFormattedDateString(long, boolean, Context)}
   */
  @Deprecated
  public static String getFormattedDateString(long utc, boolean abbreviated, Context context) {
    return TimeSince.getFormattedDateString(utc, abbreviated, context);
  }

  /**
   * @deprecated
   * Use {@link TimeSince#getFormattedDateString(long, long, boolean, Context)}
   */
  @Deprecated
  public static String getFormattedDateString(
      long start, long end, boolean abbreviated, Context context) {
    return TimeSince.getFormattedDateString(start, end, abbreviated, context);
  }
}


sample/src/main/java/com/ddiehl/timesincetextview/sample/MainActivity.java
package com.ddiehl.timesincetextview.sample;

import android.os.Bundle;
import android.support.annotation.IdRes;
import android.support.v7.app.AppCompatActivity;

import com.ddiehl.timesincetextview.TimeSinceTextView;

import java.util.Date;

import timesincetextview.sample.R;

public class MainActivity extends AppCompatActivity {
  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_main);

    long utc = new Date().getTime() / 1000;

    findView(R.id.time_now).setDate(utc);

    findView(R.id.time_seconds).setDate(utc - 30);
    findView(R.id.time_seconds_abbreviated).setDate(utc - 30);

    findView(R.id.time_minutes).setDate(utc - 5 * 60);
    findView(R.id.time_minutes_abbreviated).setDate(utc - 5 * 60);

    findView(R.id.time_hours).setDate(utc - 5 * 60 * 60);
    findView(R.id.time_hours_abbreviated).setDate(utc - 5 * 60 * 60);

    findView(R.id.time_days).setDate(utc - 3 * 24 * 60 * 60);
    findView(R.id.time_days_abbreviated).setDate(utc - 3 * 24 * 60 * 60);

    findView(R.id.time_weeks).setDate(utc - 2 * 7 * 24 * 60 * 60);
    findView(R.id.time_weeks_abbreviated).setDate(utc - 2 * 7 * 24 * 60 * 60);

    findView(R.id.time_months).setDate(utc - 6 * 30 * 24 * 60 * 60);
    findView(R.id.time_months_abbreviated).setDate(utc - 6 * 30 * 24 * 60 * 60);

    findView(R.id.time_years).setDate(utc - 4 * 52 * 7 * 24 * 60 * 60);
    findView(R.id.time_years_abbreviated).setDate(utc - 4 * 52 * 7 * 24 * 60 * 60);
  }
  
  private TimeSinceTextView findView(@IdRes int resId) {
    return (TimeSinceTextView) findViewById(resId);
  }
}


library/src/androidTest/java/com/ddiehl/timesincetextview/TimeSinceTextViewAndroidTest.java
package com.ddiehl.timesincetextview;

import android.content.Context;
import android.support.annotation.PluralsRes;
import android.support.test.InstrumentationRegistry;
import android.support.test.runner.AndroidJUnit4;

import org.hamcrest.Matcher;
import org.hamcrest.core.Is;
import org.junit.After;
import org.junit.AfterClass;
import org.junit.Before;
import org.junit.BeforeClass;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.Date;

import static org.hamcrest.MatcherAssert.assertThat;

@RunWith(AndroidJUnit4.class)
public class TimeSinceTextViewAndroidTest {
  private static Context mContext;
  private final long mTime = new Date().getTime() / 1000;

  @BeforeClass
  public static void setUpClass() {
    mContext = InstrumentationRegistry.getContext();
  }

  @AfterClass
  public static void tearDownClass() {

  }

  @Before
  public void setUp() {
  }

  @After
  public void tearDown() {
  }

  @Test
  public void getFormattedDateString_now() {
    assertThat(
        TimeSince.getFormattedDateString(
            mTime - 10, mTime, false, mContext),
        is(mContext.getString(R.string.tstv_timespan_now)));
  }

  @Test
  public void getFormattedDateString_11seconds() {
    assertThat(
        TimeSince.getFormattedDateString(
            mTime - 11, mTime, false, mContext),
        is(getQuantityString(R.plurals.tstv_timespan_seconds, 11)));
  }

  @Test
  public void getFormattedDateString_59seconds() {
    assertThat(
        TimeSince.getFormattedDateString(
            mTime - 59, mTime, false, mContext),
        is(getQuantityString(R.plurals.tstv_timespan_seconds, 59)));
  }

  @Test
  public void getFormattedDateString_1minute() {
    assertThat(
        TimeSince.getFormattedDateString(
            mTime - 1 * 60, mTime, false, mContext),
        is(getQuantityString(R.plurals.tstv_timespan_minutes, 1)));
  }

  @Test
  public void getFormattedDateString_59minutes() {
    assertThat(
        TimeSince.getFormattedDateString(
            mTime - 1 * 60 * 60 + 1, mTime, false, mContext),
        is(getQuantityString(R.plurals.tstv_timespan_minutes, 59)));
  }

  @Test
  public void getFormattedDateString_1hour() {
    assertThat(
        TimeSince.getFormattedDateString(
            mTime - 1 * 60 * 60, mTime, false, mContext),
        is(getQuantityString(R.plurals.tstv_timespan_hours, 1)));
  }

  @Test
  public void getFormattedDateString_23hours() {
    assertThat(
        TimeSince.getFormattedDateString(
            mTime - 1 * 24 * 60 * 60 + 1, mTime, false, mContext),
        is(getQuantityString(R.plurals.tstv_timespan_hours, 23)));
  }

  @Test
  public void getFormattedDateString_1day() {
    assertThat(
        TimeSince.getFormattedDateString(
            mTime - 1 * 24 * 60 * 60, mTime, false, mContext),
        is(getQuantityString(R.plurals.tstv_timespan_days, 1)));
  }

  @Test
  public void getFormattedDateString_6days() {
    assertThat(
        TimeSince.getFormattedDateString(
            mTime - 1 * 7 * 24 * 60 * 60 + 1, mTime, false, mContext),
        is(getQuantityString(R.plurals.tstv_timespan_days, 6)));
  }

  @Test
  public void getFormattedDateString_1week() {
    assertThat(
        TimeSince.getFormattedDateString(
            mTime - 1 * 7 * 24 * 60 * 60, mTime, false, mContext),
        is(getQuantityString(R.plurals.tstv_timespan_weeks, 1)));
  }

  @Test
  public void getFormattedDateString_4weeks() {
    assertThat(
        TimeSince.getFormattedDateString(
            mTime - 1 * 30 * 24 * 60 * 60 + 1, mTime, false, mContext),
        is(getQuantityString(R.plurals.tstv_timespan_weeks, 4)));
  }

  @Test
  public void getFormattedDateString_1month() {
    assertThat(
        TimeSince.getFormattedDateString(
            mTime - 1 * 30 * 24 * 60 * 60, mTime, false, mContext),
        is(getQuantityString(R.plurals.tstv_timespan_months, 1)));
  }

  @Test
  public void getFormattedDateString_12months() {
    assertThat(
        TimeSince.getFormattedDateString(
            mTime - 1 * 365 * 24 * 60 * 60 + 1, mTime, false, mContext),
        is(getQuantityString(R.plurals.tstv_timespan_months, 12)));
  }

  @Test
  public void getFormattedDateString_1year() {
    assertThat(
        TimeSince.getFormattedDateString(
            mTime - 1 * 365 * 24 * 60 * 60, mTime, false, mContext),
        is(getQuantityString(R.plurals.tstv_timespan_years, 1)));
  }

  @Test
  public void getFormattedDateString_30years() {
    assertThat(
        TimeSince.getFormattedDateString(
            mTime - 30 * 365 * 24 * 60 * 60, mTime, false, mContext),
        is(getQuantityString(R.plurals.tstv_timespan_years, 30)));
  }

  @Test
  public void getFormattedDateString_abbrev() {
    assertThat(
        TimeSince.getFormattedDateString(
            mTime - 59, mTime, true, mContext),
        is(getQuantityString(R.plurals.tstv_timespan_seconds_abbr, 59)));
    assertThat(
        TimeSince.getFormattedDateString(
            mTime - 1 * 60, mTime, true, mContext),
        is(getQuantityString(R.plurals.tstv_timespan_minutes_abbr, 1)));
    assertThat(
        TimeSince.getFormattedDateString(
            mTime - 1 * 60 * 60, mTime, true, mContext),
        is(getQuantityString(R.plurals.tstv_timespan_hours_abbr, 1)));
    assertThat(
        TimeSince.getFormattedDateString(
            mTime - 1 * 24 * 60 * 60, mTime, true, mContext),
        is(getQuantityString(R.plurals.tstv_timespan_days_abbr, 1)));
    assertThat(
        TimeSince.getFormattedDateString(
            mTime - 1 * 7 * 24 * 60 * 60, mTime, true, mContext),
        is(getQuantityString(R.plurals.tstv_timespan_weeks_abbr, 1)));
    assertThat(
        TimeSince.getFormattedDateString(
            mTime - 1 * 30 * 24 * 60 * 60, mTime, true, mContext),
        is(getQuantityString(R.plurals.tstv_timespan_months_abbr, 1)));
    assertThat(
        TimeSince.getFormattedDateString(
            mTime - 1 * 365 * 24 * 60 * 60, mTime, true, mContext),
        is(getQuantityString(R.plurals.tstv_timespan_years_abbr, 1)));
  }

  @Test
  public void setDate_long() {
    TimeSinceTextView view = new TimeSinceTextView(mContext);
    view.setDate(mTime - 1 * 24 * 60 * 60); // 1 day
    assertThat(
        view.getText().toString(),
        is(getQuantityString(R.plurals.tstv_timespan_days, 1)));
  }

  @Test
  public void setDate_java_util_date() {
    TimeSinceTextView view = new TimeSinceTextView(mContext);
    view.setDate(new Date((mTime - 1 * 24 * 60 * 60) * 1000)); // 1 day
    assertThat(
        view.getText().toString(),
        is(getQuantityString(R.plurals.tstv_timespan_days, 1)));
  }

  private static Matcher<String> is(String matcher, Object... args) {
    return Is.is(
        String.format(matcher, args));
  }

  private String getQuantityString(@PluralsRes int resId, int n) {
    return mContext.getResources().getQuantityString(resId, n, n);
  }
}
